meld/cluster.py

```diff
@@ -272,11 +272,27 @@ def sparse_dim_connectivity(dim_con):
             # extend the row/col connections
             rows.extend(np.nonzero(dind[i] == r[j])[0])
             cols.extend(np.nonzero(dind[i] == c[j])[0])
-
-    # create the sparse connectivity matrix
-    data = np.ones(len(rows))
-    cmat = sparse.coo_matrix((data, (rows, cols)),
-                             shape=(nelements, nelements))
+    if (not triu) | row_dict:
+        tmp = rows.copy()
+        rows.extend(cols)
+        cols.extend(tmp)
+
+    if row_dict:
+        cmat = {}
+        for r, c in zip(rows, cols):
+            try:
+                cmat[r] = np.append(cmat[r], c)
+            except KeyError:
+                cmat[r] = np.array([c])
+    elif row_list:
+        cmat = [np.array([], dtype=int) for r in rows]
+        for r, c in zip(rows, cols):
+            cmat[r] = np.append(cmat[r], int(c))
```

Comment:
This row_list version looks like it is what I suggested, and I would expect it to be the fastest once you are inside the tfce_pos main loop.  However, I'm not sure where this function gets called (hopefully it is called outside whatever loop iterates over maps in your file(s)).

It should be noted that tfce_pos will only access each node's connectedness once, and therefore only twice per map if you do both positives and negatives, so for this to be effective, the data structure for connectedness should be set up and saved, and reused by a large number of tfce_pos calls (without copying it, if possible).

Reply:
This does get called external to the tfce_pos function. Just a single connectivity object that doesn't have to get copied around.